pkg/apis/core/validation/shoot.go

```diff
@@ -1029,6 +1038,88 @@ func validateKubeletConfigEvictionSoftGracePeriod(eviction *core.KubeletConfigEv
 	return allErrs
 }
 
+// Reserved PIDs have to be non-negative
+const ReservedPIDsMinimum int64 = 0
+
+func validateKubeletConfigKubeReserved(reserved *core.KubeletConfigKubeReserved, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if reserved.CPU != nil {
+		allErrs = append(allErrs, validateResourceQuantityValue("cpu", *reserved.CPU, fldPath.Child("cpu"))...)
+	}
+	if reserved.Memory != nil {
+		allErrs = append(allErrs, validateResourceQuantityValue("memory", *reserved.Memory, fldPath.Child("memory"))...)
+	}
+	if reserved.EphemeralStorage != nil {
+		allErrs = append(allErrs, validateResourceQuantityValue("ephemeralStorage", *reserved.EphemeralStorage, fldPath.Child("ephemeralStorage"))...)
+	}
+	if *reserved.PID < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("pid"), *reserved.PID, fmt.Sprintf("PID must be atleast %d", ReservedPIDsMinimum)))
+	}
+	return allErrs
+}
+
+func validateKubeletConfigSystemReserved(reserved *core.KubeletConfigSystemReserved, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if reserved.CPU != nil {
+		allErrs = append(allErrs, validateResourceQuantityValue("cpu", *reserved.CPU, fldPath.Child("cpu"))...)
+	}
+	if reserved.Memory != nil {
+		allErrs = append(allErrs, validateResourceQuantityValue("memory", *reserved.Memory, fldPath.Child("memory"))...)
+	}
+	if reserved.EphemeralStorage != nil {
+		allErrs = append(allErrs, validateResourceQuantityValue("ephemeralStorage", *reserved.EphemeralStorage, fldPath.Child("ephemeralStorage"))...)
+	}
+	if *reserved.PID < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("pid"), *reserved.PID, fmt.Sprintf("PID must be atleast %d", ReservedPIDsMinimum)))
+	}
+	return allErrs
+}
+
+const KubeReservedAllocatableName string = "kube-reserved"
+const SystemReservedAllocatableName string = "system-reserved"
+const PodsAllocatableName string = "pods"
+const KubeReservedCgroupName string = "KubeReservedCgroupName"
+const systemReservedCgroupName string = "SystemReservedCgroupName"
+
+func validateKubeletConfigEnforceNodeAllocatable(nodeAllocatable []string, kubeReservedCgroup, systemReservedCgroup *string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	var (
+		enforcePods   bool
+		enforceKube   bool
+		enforceSystem bool
+	)
+	if len(nodeAllocatable) > 0 {
+		for idx := range nodeAllocatable {
+			switch allocatable := nodeAllocatable[idx]; allocatable {
+			case PodsAllocatableName:
+				if enforcePods {
+					allErrs = append(allErrs, field.Invalid(fldPath.Index(idx), allocatable, "Enforced Node Allocatable should not be repeated"))
```

Comment:
```suggestion
					allErrs = append(allErrs, field.Duplicate(fldPath.Index(idx), PodsAllocatableName))
```
Also for the duplicate value checks below.